### PR TITLE
[WIP] Enable double dollar for triple quoted interpolated strings

### PR DESCRIPTION
**Pull request by alfonsogarciacaro**
See https://github.com/fsharp/fslang-suggestions/issues/1150

This doesn't include percent symbol yet, I'm sending the draft PR to see if the approach is correct. I started trying to implement variable dollar length as in C# but then I realized `LexerStringKind` is byte encoded in ServiceLexing.fs, so I kept is as a flag for simplicity. If we finally only allow double dollar (and not triple, quadruple) this should be stated in the RFC.

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/fsharp?pullRequest=13259) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/fsharp?pullRequest=13259) before we can accept your contribution.<br/><br/>:x: alfonsogarciacaro [sign now](https://cla.dotnetfoundation.org/dotnet/fsharp?pullRequest=13259)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/fsharp?pullRequest=13259) it.</sub>

File: src/Compiler/lex.fsl [hunk 1]
```diff
@@ -845,41 +846,43 @@ rule token args skip = parse
...
+       | _ -> ()
        // To continue token-by-token lexing may involve picking up the new args.stringNes
        let cont = LexCont.Token(args.ifdefStack, args.stringNest)
        LBRACE cont
      }
 
  | "|" { BAR }
 
- | "}"
+ | ("}}" | "}")
```
Comment (dsyme) [inline src/Compiler/lex.fsl:865]: This is parsing "}}" as a single "}" in regular F# code, hence the tests failing to compile

Comment (dsyme) [reply src/Compiler/lex.fsl:865]: You can probably add a new RBRACE_RBRACE token and smash it apart in the lexfilter. Though be careful of |}} too and check if anything else can be combined with }. 

Likewise LBRACE of course

Comment (dsyme) [reply src/Compiler/lex.fsl:865]: Sample smash is here:

https://github.com/dotnet/fsharp/blob/main/src/Compiler/SyntaxTree/LexFilter.fs#L2442

Not sure there's any better way to do it

Comment (alfonsogarciacaro) [reply src/Compiler/lex.fsl:868]: @dsyme As I don't know well how the lexfilter works and as you say, we can have problems with things like `{{|`, I'm trying to [move back the lexeme end position](https://github.com/dotnet/fsharp/pull/13259/commits/57e81fb713c88a44fd029adbcabda12037c312c7) if we shouldn't consume the two braces. Is this a bad idea?

Also, for some reason CI is not running tests. I cannot get repo to build locally either. When I run `Build.cmd` at one point I get these error (same in `main` branch):

```
C:\Users\alfon\repos\fsharp\tests\FSharp.Core.UnitTests\FSharp.Core\RefStructs.fs(20,18): error FS0412: A type instanti
ation involves a byref type. This is not permitted by the rules of Common IL. [C:\Users\alfon\repos\fsharp\tests\FSharp
.Core.UnitTests\FSharp.Core.UnitTests.fsproj]
C:\Users\alfon\repos\fsharp\tests\FSharp.Core.UnitTests\FSharp.Core\RefStructs.fs(20,6): error FS0437: A type would sto
re a byref typed value. This is not permitted by Common IL. [C:\Users\alfon\repos\fsharp\tests\FSharp.Core.UnitTests\FS
harp.Core.UnitTests.fsproj]
```

Comment (vzarytovskii) [reply src/Compiler/lex.fsl:868]: CI is not running tests if there's a merge conflict, IIRC.
The local issue is weird, does this happen if you build with `-noVisualStudio` switch too?

Comment (alfonsogarciacaro) [reply src/Compiler/lex.fsl:868]: Thanks a lot @vzarytovskii! I will fix the merge conflicts 👍 And the `-noVisualStudio` switch does indeed work! I do have Visual Studio installed but I don't use it a lot and maybe the installation is faulty (I had problems to open FSharp.sln with VS).

Comment (vzarytovskii) [reply src/Compiler/lex.fsl:868]: > Thanks a lot @vzarytovskii! I will fix the merge conflicts 👍 And the `-noVisualStudio` switch does indeed work! I do have Visual Studio installed but I don't use it a lot and maybe the installation is faulty (I had problems to open FSharp.sln with VS).

Huh, interesting, thanks. I will add some notes to the DEVGUIDE on how to deal with such errors (incl. VS repair and building with the flag).

File: src/Compiler/lex.fsl [hunk 2]
```diff
@@ -854,6 +854,9 @@ rule token args skip = parse
          // Note, we do not update the 'm', any incomplete-interpolation error
          // will be reported w.r.t. the first '{'
          args.stringNest <- (counter + 1, style, isDoubleDollar, m) :: rest
+       | _ when isDouble ->
+           lexbuf.LexemeLength <- 1
```
[the remark this answers is not in the document]
Comment (alfonsogarciacaro) [reply src/Compiler/lex.fsl:858]: @dsyme Maybe, I'm not sure. Tests are failing but it seems that it's not detecting well when an interpolation gap closes now. I wasn't sure how to debug the lexer tests. Unfortunately I won't be able to work on the PR this week, so it'd be great if someone could take over.

**Comment by alfonsogarciacaro** [comment]
Sorry, have to close this PR because I couldn't make the parser work. If someone with better knowledge can take it over it'd be great, as this feature will be very beneficial when embedding HTML, CSS, JSON, etc in F#.

**Pull request closed without merging**